File: cs/src/AutoDiff/ConstPower.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;

namespace AutoDiff
{
    /// <summary>
    /// Represents a constant-power function x^n, where n is constant.
    /// </summary>
    [Serializable]
    public class ConstPower : Term
    {
        /// <summary>
        /// Constructs a new instance of the <see cref="ConstPower"/> class.
        /// </summary>
        /// <param name="baseTerm">The base of the power function</param>
        /// <param name="exponent">The exponent of the power function</param>
        public ConstPower(Term baseTerm, double exponent)
        {
            Base = baseTerm;
            Exponent = exponent;
        }

        /// <summary>
        /// Gets the base term of the power function
        /// </summary>
        public Term Base { get; private set; }

        /// <summary>
        /// Gets the exponent term of the power function.
        /// </summary>
        public double Exponent { get; private set; }

        /// <summary>
        /// Accepts a term visitor.
        /// </summary>
        /// <param name="visitor">The term visitor to accept.</param>
        public override void Accept(ITermVisitor visitor)
        {
            visitor.Visit(this);
        }

        /// <summary>
        /// Accepts a term visitor with a generic result
        /// </summary>
        /// <typeparam name="TResult">The type of the result from the visitor's function</typeparam>
        /// <param name="visitor">The visitor to accept</param>
        /// <returns>
        /// The result from the visitor's visit function.
        /// </returns>
        public override TResult Accept<TResult>(ITermVisitor<TResult> visitor)
        {
            return visitor.Visit(this);
        }
		public override string ToString()
		{
			return string.Format("constPower( {0}, {1} )", Base, Exponent);
		}
		
		public override Term AggregateConstants ()
		{
			Base = Base.AggregateConstants();
			if (Base is Constant) {
				return Math.Pow ((Base as Constant).Value, Exponent);
			} else if (Base is Zero) {
				if (Exponent >= 0) {
					return Base;
				} else {
					throw new DivideByZeroException ();
				}
			} else if (Base is ConstPower) {
				Exponent *= (Base as ConstPower).Exponent;
				Base = (Base as ConstPower).Base;
				return this;
			  }else {
				return this;
			}
		}
		public override Term Derivative(Variable v)
		{
			return this.Exponent * new ConstPower(this.Base,(this.Exponent-1)) * this.Base.Derivative(v);
		}		
    }
}

File: java/src/main/java/AutoDiff/ConstPower.java
//
// Translated by CS2J (http://www.cs2j.com): 12.11.2016 09:40:33
//

package AutoDiff;

/**
* Represents a constant-power function x^n, where n is constant.
*/
public class ConstPower  extends Term 
{
    /**
    * Constructs a new instance of the 
    *  {@link #ConstPower}
    *  class.
    * 
    *  @param baseTerm The base of the power function
    *  @param exponent The exponent of the power function
    */
    public ConstPower(Term baseTerm, double exponent) throws Exception {
        setBase(baseTerm);
        setExponent(exponent);
    }

    /**
    * Gets the base term of the power function
    */
    private Term Base;
    public Term getBase() {
        return Base;
    }

    public void setBase(Term value) {
        Base = value;
    }

    /**
    * Gets the exponent term of the power function.
    */
    private double Exponent;
    public double getExponent() {
        return Exponent;
    }

    public void setExponent(double value) {
        Exponent = value;
    }

//    /**
//    * Accepts a term visitor.
//    *
//    *  @param visitor The term visitor to accept.
//    */
//    public void accept(ITermVisitor visitor) throws Exception {
//        visitor.visit(this);
//    }

    /**
    * Accepts a term visitor with a generic result
    * The type of the result from the visitor's function
    *  @param visitor The visitor to accept
    *  @return 
    * The result from the visitor's visit function.
    */
    public <TResult>TResult accept(ITermVisitor<TResult> visitor) throws Exception {
        return visitor.visit(this);
    }

    public String toString() {
        try
        {
            return String.format("constPower( %s, %f )", getBase(), getExponent());
        }
        catch (RuntimeException exception)
        {
            throw exception;
        }
        catch (Exception exception)
        {
            throw new RuntimeException(exception);
        }
    
    }

    public Term aggregateConstants() throws Exception {
        setBase(getBase().aggregateConstants());
        if (getBase() instanceof Constant)
        {
            return TermBuilder.Constant(Math.pow((getBase() instanceof Constant ? (Constant)getBase() : (Constant)null)
                .getValue(),
                getExponent()));
        }
        else if (getBase() instanceof Zero)
        {
            if (getExponent() >= 0)
            {
                return getBase();
            }
            else
            {
                throw new IllegalArgumentException("Argument 'divisor' is 0");
            } 
        }
        else if (getBase() instanceof ConstPower)
        {
            setExponent(getExponent() * (getBase() instanceof ConstPower ? (ConstPower)getBase() : (ConstPower)null).getExponent());
            setBase((getBase() instanceof ConstPower ? (ConstPower)getBase() : (ConstPower)null).getBase());
            return this;
        }
        else
        {
            return this;
        }   
    }

    public Term derivative(Variable v) throws Exception {
        return Term.multiply(TermBuilder.Constant(this.getExponent()), Term.multiply(new ConstPower(this.getBase(),(this
            .getExponent() - 1)), this.getBase().derivative(v)));
    }

}


